git: add jj git sync command

0xdeafbeef opened the pull request:
<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the commits rather than adding commits on top. Use force-push when
pushing the updated commits (`jj git push` does that automatically when you
rewrite commits). Merge the PR at will once it's been approved. See
https://github.com/jj-vcs/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->

# Checklist

If applicable:

- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [x] I have added/updated tests to cover my changes

Closes #1039

0xdeafbeef commented:
Wrote it according https://github.com/jj-vcs/jj/issues/1039#issuecomment-2465742400.

PhilipMetzger reviewed (commented):
🎉 finally, here are some minor nits.

yuja commented on cli/src/commands/git/sync.rs:
```diff
@@ -0,0 +1,261 @@
+    let mut pre_fetch_heads: HashMap<RemoteRefSymbolBuf, CommitId> = HashMap::new();
+
+    for remote in &remotes {
+        for (name, remote_ref) in tx.repo().view().remote_bookmarks(remote) {
+            // Only capture non-conflicted targets
+            if let Some(commit_id) = remote_ref.target.as_normal() {
```
I think we need to use the local bookmark tracking the remote. Remote bookmarks often point to hidden commits.

0xdeafbeef commented:
Status report - i'll try to finish it on this weekend. Daily driving it, found a nasty bug :)

0xdeafbeef commented:
@yuja 
> I think we need to use the local bookmark tracking the remote. Remote bookmarks often point to hidden commits.

It didn't work well, so i've a hybrid solution.

The problem which i debuged for so long was when 2 bookmarks pointed to the same comit and second bookmark moved in remote. Sync will move my local bookmark to the same position.

0xdeafbeef commented on cli/tests/test_git_sync.rs:
```diff
@@ -0,0 +1,403 @@
+// Copyright 2020-2023 The Jujutsu Authors
```
done

jgilchrist commented:
@0xdeafbeef I've tried this out and unless I'm misunderstanding, it doesn't seem to be working as expected. Prior to running `jj git sync` things look like:

<img width="795" height="200" alt="image" src="https://github.com/user-attachments/assets/04e30200-ae89-4343-ba03-cfe2ef8974a4" />

i.e. I've got a few changes, some with bookmarks, all based on `main`. `main@upstream` has a single new commit, so I'm expecting that after the sync, all of these changes are rebased. However, after `jj git sync`, I see:

<img width="1485" height="114" alt="image" src="https://github.com/user-attachments/assets/cc074dee-8e9d-4f00-bae8-2cfaa99b4dd4" />

i.e. all of the WIP changes are hidden. The output of the sync shows:

<img width="1018" height="140" alt="image" src="https://github.com/user-attachments/assets/b669777f-e3c0-4434-a515-226b14b7079c" />

It looks like we do attempt to rebase the changes ('Rebasing 7 commits') but then they are somehow not rebased successfully ('7 already merged').

Any idea what might be going wrong here?

0xdeafbeef commented:
> Please don't let this PR become stale! I'm super excited to see it merged - is there anything you need in terms of testing/data to get this over the line @0xdeafbeef? 🙏

Example of @jgilchrist showed that my impl is unsound and works only in basic cases. I'll try to fix it in the next 2 days and push updated version. Sync won't have merge support cause this rabbit hole appears to be deep enough :)

PhilipMetzger reviewed (commented):
SGTM but Yuya will do the final reviews

yuja commented:
> Now we use fork-point rebasing for each changed bookmark, equivalent to `jj rebase --branch bookmark --destination bookmark@remote`

I haven't reviewed the implementation thoroughly, but does it rebase anonymous branches?
```
                      B'
B                     |
|               =>    A' foo@origin
A foo@origin       A
```

The condition to rebase might be something like `old_base..head & ::untracked_remote_bookmarks() == none()`.

BTW, I expect the core logic and tests will be moved to jj-lib.

0xdeafbeef commented:
> > @0xdeafbeef I'll refactor and rebase onto main after you approve/disapprove design changes
> 
> Is there anything you need from the `jj` maintainers to get this merged, or is it simply a matter of finding the time to work on it? 🙏

problem is the lack of time and that it's not so trivial as i thought :) I'll rebase on main today and try to resolve all comments

zx8 commented:
I think it's clear at this point that @0xdeafbeef doesn't have the time to finish this PR up. This isn't a complaint, I understand how life works, just an observation. Especially considering that I don't have the skillset to help carry it over the line myself!

That said, it's been open for a while now and it's one of the few remaining gaps that gives me pause before recommending `jj` to people. Is there any way I or others could help move it forward, whether that's testing, picking up specific pieces, or someone else stepping in to finish it up?

0xdeafbeef commented:
> I think it's clear at this point that @0xdeafbeef doesn't have the time to finish this PR up. This isn't a complaint, I understand how life works, just an observation. Especially considering that I don't have the skillset to help carry it over the line myself!
> 
> That said, it's been open for a while now and it's one of the few remaining gaps that gives me pause before recommending `jj` to people. Is there any way I or others could help move it forward, whether that's testing, picking up specific pieces, or someone else stepping in to finish it up?

I've used jj sync for a while and found it kinda useless :( For big repos sync takes really LONG time. Eg for linux it will try to rebase every tag/branch onto upstream and it's wasteful. And also I'm not sure that the rebase of all braches is desired and expected behavior. Personally i just do `jj git fetch && jj rebase --into branch@remote`. I have time now and i can recheck the status(I guess current should be reimpled from scratch). But code itself is much simpler than ergonomics side. Mb we should discuss it in Discord(or here) 1 more time. And yeah, sorry for going radio silent :)

martinvonz commented:
IMO, `jj sync` should default to rebasing `reachable(@, trunk()..)` and have a `--all` switch to rebase all revisions.

matts1 commented:
> For big repos sync takes really LONG time

If you're rebasing tags, that sounds like you've either got it rebasing immutable commits on sync, which it definitely shouldn't do, or for some reason your tags are mutable (which seems like misconfiguration to me). I have jj configured to treat only my local branches as mutable, so thus rebasing "everything" is just rebasing a small set of mutable commits.

The way I've been syncing is to use `jj git fetch && jj rebase -s 'roots(mutable())' -d 'trunk()'` (or equivalently, `-b 'mutable()'). I've been doing this on chromium, which is also a massive repository, and the rebase is an extremely quick operation.

See https://source.chromium.org/chromium/chromium/src/+/main:tools/jj/sync.py;drc=dbb5f40c731c7171b295d62b96463d56c6e6bbc1

0xdeafbeef committed:
git: add planner for syncing selected commits after import

Add the jj-lib plumbing for a future `jj git sync` command, providing
pull-like behavior similar to `git pull --rebase`.

The planner rebases selected local commits onto fast-forwarded upstream
positions while preserving selected branches and merges. Multiple independent
upstream movements are planned and applied as a single rewrite operation.

The planner is deliberately conservative. It handles only unambiguous
fast-forward movements relevant to the selection. It rejects deleted,
conflicted, or non-fast-forward remote updates; incompatible destinations or
parent replacements; selections that cross unselected intermediate commits;
conflicting rewrite roles; and plans that would rewrite an imported
destination.

0xdeafbeef commented:
@zx8 finally i've found the time. Rewrote sync from scratch, for now only lib part is done.

It's much much trickier than it seams. Will write a bigger comment a bit latter :)

UPD:

It should cover normal use cases and fail fast on strange graph shapes. Played with Z3 (thanks, we have LLMs) to check that the graph isn't broken after the rebase. Found a few bugs using it, and this version is much more robust than the previous one.

What it does and does not do:

We select all commits from the revsets describing the local branch. Then we find the upstream point from which each child branch started and rebase it onto the new upstream position.

We skip cases where:
- upstream was deleted, conflicted or force-pushed;
- different upstream updates would move the same branch in different directions;
- there are unselected commits in the middle of the branch;
- several updates overlap in a way that could lose commits or graph edges.

Some tricky things that should work:
- anonymous branches without bookmarks;
- branches containing merges;
- several independent upstream branches moving at once;
- bookmarks and working copies following the rewritten commits;
- unrelated branches staying untouched.

For now only the lib part and tests are done. CLI will come later. And it should be pretty it easy to add.

I'll make it draft for now, need a bit more time  to sort everything out  before asking for a review

0xdeafbeef commented:
@yuja @martinvonz

i've found another design issue (i've lost count of them) :sob: :sob: :sob: :sob: :sob:

Right now the planner consumes the output of git import after fetch. From that output, it knows that a remote bookmark moved from A to B during the fetch.

This makes it possible to find local branches based on A and rebase them onto B. It also works for anonymous branches, which is pretty important for a normal jj workflow.

But it makes the CLI a bit weird:

```sh
jj git sync
```

can fetch and rebase the branch, while:

```sh
jj git fetch
jj git sync
```

makes the second command a no-op, because the A -> B movement was only available during the previous import.

I can see three possible options:

1. Keep the current design and accept that sync only works together with its own fetch. A separate fetch does not preserve the information needed by a later sync.

2. Try to infer the upstream only from the current commit graph. I don't think this is safe, and I tried it in the first iteration:

```text
              C  (@, anonymous branch)
             /
    A --- P --- R --- D  (release@origin)
```

The graph says that rebasing C onto D is valid:

```text
    A --- P --- R --- D --- C'
```

But it does not say whether C was actually created from `release@origin` or is an independent branch that should stay on P. Both interpretations have exactly the same graph, so ancestry alone is not enough.

3. Persist a small, non-consumable record of imported movement for each tracked remote bookmark.

For example, before fetch there may be two anonymous branches based on A:

```text
A (main@origin) ──┬──> C (@)
                  └──> E (another anonymous head)
```

Fetch imports A -> B:

```text
A ──┬──> B (main@origin)
    ├──> C (@)
    └──> E (another anonymous head)

stored sync frontier: main@origin: A -> B
```

A later, separate sync can rebase only the current branch:

```text
A ──┬──> B (main@origin) ──> C' (@)
    └──> E (another anonymous head)

stored sync frontier: main@origin: A -> B
```

The frontier is not consumed, so a later `--all` can still rebase E:

```text
A ──> B (main@origin) ──┬──> C' (@)
                        └──> E'

stored sync frontier: main@origin: A -> B
```

Sync would only consider stored movements matching the remote and bookmark scope selected by the command.

Commits already based on the current target will naturally be skipped by the next sync.

There are some annoying details here. If fetch deletes or force-pushes the bookmark, a later sync still needs to know about it and fail. But push, track/untrack, or a manual remote bookmark change should probably forget the saved movement, since it no longer describes a fetch.

This does not require assigning an upstream to every anonymous branch. The revset still selects the commits, the persisted frontier only tells the planner which remote bookmarks moved, and the existing graph checks decide what can be safely rebased.

tl;dr: `View` already stores the current remote target B. I think it also needs to remember that the imported movement started at A, or the first unsafe update if the bookmark was deleted or force-pushed. Sync uses that state but does not consume it.

This needs some merge/serialization semantics, but it seems less surprising than making sync depend on whether fetch happened in the same command.

Does this sound reasonable, or is there some existing state/API that would make this simpler?

It’s the deepest rabbit hole I’ve ever mistaken for an easy weekend pr :)

PhilipMetzger commented on lib/src/git_sync.rs:
```diff
@@ -0,0 +1,628 @@
+}
+
+/// Why an imported remote bookmark update cannot be synchronized.
+#[expect(missing_docs)]
```
nit: write the documentation for new modules. You also had a previous version which already dealt with this, with a minor nit that the doc comments should be above all `#[derive(...)]` or `#[error(...)` annotations.

0xdeafbeef replied:
I thought the documenting error variants which have literal text is a bit dumb, but added docs

PhilipMetzger commented on lib/src/git_sync.rs:
```diff
@@ -0,0 +1,628 @@
+}
+
+/// Errors from synchronizing selected commits with imported Git updates.
+#[expect(missing_docs)]
```
ditto

PhilipMetzger commented on lib/src/git_sync.rs:
```diff
@@ -0,0 +1,628 @@
+                affected_commits: IndexSet::new(),
+                rewrite_roots: IndexSet::new(),
+            })
+            .collect();
```
nit: use `collect_vec` to adhere to our style guide.

PhilipMetzger commented on lib/src/git_sync.rs:
```diff
@@ -0,0 +1,628 @@
+                .affected_commits
+                .difference(&reachable)
+                .cloned()
+                .collect();
```
ditto with `collect_vec()`

PhilipMetzger commented on lib/src/git_sync.rs:
```diff
@@ -0,0 +1,628 @@
+            .into_iter()
+            .map(|selection| selection.boundary.new)
+            .filter(|id| planned_ids.contains(id))
+            .collect();
```
ditto with `collect_vec()`

PhilipMetzger commented on lib/src/git_sync.rs:
```diff
@@ -0,0 +1,628 @@
+            .into_iter()
+            .map(|commit| (commit.id().clone(), commit))
+            .collect::<IndexMap<_, _>>();
+        let mut children_by_parent = IndexMap::<CommitId, Vec<CommitId>>::new();
```
optional: I'm not sure but I think the type can be inferred here so it just can be `IndexMap::new`

PhilipMetzger commented on lib/src/git_sync.rs:
```diff
@@ -0,0 +1,628 @@
+                        .get_remote_bookmark(update.symbol.as_ref())
+                        .is_tracked()
+            })
+            .collect();
```
ditto with `collect_vec` and dropping the annotation.

martinvonz commented:
> Keep the current design and accept that sync only works together with its own fetch. A separate fetch does not preserve the information needed by a later sync.

This sounds good to me.

PhilipMetzger commented on lib/src/repo.rs:
```diff
@@ -1442,6 +1442,83 @@ impl MutableRepo {
+    /// The caller must discard the transaction if this method returns an error
+    /// or if its future is cancelled before completion. On failure, the
+    /// repository may contain partial in-memory mutations.
+    #[cfg(feature = "git")]
```
nit: I don't think this needs to be Git only

PhilipMetzger commented on lib/src/git_sync.rs:
```diff
@@ -0,0 +1,628 @@
+            .await?;
+        Ok(rebased_commits)
+    }
+}
+
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+struct BoundaryMove {
+    old: CommitId,
+    new: CommitId,
+}
+
+struct BoundarySelection {
+    boundary: BoundaryMove,
+    affected_commits: IndexSet<CommitId>,
+    rewrite_roots: IndexSet<CommitId>,
+}
+
+struct GitSyncPlanBuilder<'repo> {
+    repo: &'repo MutableRepo,
+    selected: SelectedGraph,
+    boundaries: Vec<BoundarySelection>,
+    parent_replacements: IndexMap<CommitId, IndexMap<CommitId, CommitId>>,
+}
+
```
nit: I guess these types should have some internal documentation

PhilipMetzger commented on lib/src/git_sync.rs:
```diff
@@ -0,0 +1,628 @@
+    }
+
+    async fn classify_affected_commits_and_rewrite_roots(&mut self) -> Result<(), GitSyncError> {
+        let index = self.repo.index();
```
nit: I'd be nice if the algorithm here had a explanation.

PhilipMetzger reviewed (commented):
minor comments. Also will you add the final implementation commit for `jj git sync` here as a parent or will that be a separate PR?

0xdeafbeef commented:
> minor comments. Also will you add the final implementation commit for `jj git sync` here as a parent or will that be a separate PR?

I've already started implementing the cli as a second commit in this PR, and that's how I found the fetch + sync issue described above. I don't think it makes sense to split them into separate PRs, cause cli shapes lib after all
